colossalai/nn/optimizer/distributed_came.py

```diff
@@ -0,0 +1,284 @@
+        r_factor = (exp_avg_sq_row / exp_avg_sq_row_mean).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+
+        return torch.mul(r_factor, c_factor)
+
+    def _unflatten_grad_tensor_by_param(self, param):
+        ori_shape = self.ori_shape[id(param)]
+        self.working_shape[id(param)]
+        # return param.grad.data.reshape(*working_shape)
+        if not (len(ori_shape) >= 2 and len(param.grad.data.shape) == 1):
+            return param.grad.data
+        remaining_dims = ori_shape[1:]
+        if param.grad.data.shape[0] % torch.prod(torch.tensor(remaining_dims)) != 0:
+            self.gather_before_compute[id(param)] = True
+            gathered_grad = [
+                torch.zeros_like(param.grad.data) for _ in range(dist.get_world_size(group=self.zero_parallel_group))
+            ]
+            dist.all_gather(gathered_grad, param.grad.data, group=self.zero_parallel_group)
+            gathered_grad = torch.cat(gathered_grad, dim=-1).reshape(*ori_shape)
+            assert gathered_grad.shape == ori_shape
+            return gathered_grad
+        return param.grad.data.reshape(-1, *remaining_dims)
+
+    def _flatten_update_tensor_by_param(self, param, tensor):
+        self.working_shape[id(param)]
+        # return tensor.reshape(*working_shape)
+        ori_shape = self.ori_shape[id(param)]
+        if not (len(ori_shape) >= 2 and len(param.grad.data.shape) == 1):
+            return tensor
+        if self.gather_before_compute[id(param)]:
+            rank = dist.get_rank(group=self.zero_parallel_group)
+            length = param.data.shape[0]
+            return torch.flatten(tensor)[rank * length : (rank + 1) * length]
+        return torch.flatten(tensor)
+
+    def step(self, closure=None):
+        """Performs a single optimization step.
+        Args:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = self._unflatten_grad_tensor_by_param(p) if self.zero else p.grad.data
+                if grad.dtype in {torch.float16, torch.bfloat16}:
+                    grad = grad.float()
+                if grad.is_sparse:
+                    raise RuntimeError("CAME does not support sparse gradients.")
+
+                state = self.state[p]
+                # zero下grad_shape是原始grad经过flatten后再切割(只有一维)
+                grad_shape = grad.shape
+
+                factored = self._get_options(grad_shape)
+                # State Initialization
+                if len(state) == 0:
+                    state["step"] = 0
+
+                    state["exp_avg"] = torch.zeros_like(grad)
```

Comment:
zeros_like(p) is better

Reply:
I will fix this, but why zeros_like(p) is better?

Reply:
Cause master weight always has the same device and dtype with optimizer states